### ISSUE TITLE
failed to produce bindings for RPM library

### ISSUE THREAD

**Issue by sa2ajj** [rust-lang-nursery/rust-bindgen#363]
I'm looking into creating Rust bindings for the RPM libraries.

First I installed `bindgen`, but it seems to be somewhat inconsistent with what `servo/rust-bindgen` offers.  So my second attempt was to clone and build `bindgen` from `servo/rust-bindgen`:

```
$ git clone https://github.com/servo/rust-bindgen
$ cd rust-bindgen/bindgen
$ cargo build --release
```

after that I created a simple `wrapper.h`:
```
#include <rpm/rpmlib.h>
#include <rpm/rpmts.h>
```

and then run `bindgen` on it:
```
$ CLANG_PATH=/opt/clang+llvm-3.9.1-x86_64-linux-gnu-debian8/bin/clangLIBCLANG_PATH=/opt/clang+llvm-3.9.1-x86_64-linux-gnu-debian8/lib/ ~/bin/bindgen wrapper.h --output=rpm.rs
```

as the result I got these errors:
```
ERROR:libbindgen::ir::context: Valid declaration with no USR: Cursor(__va_list_tag kind: StructDecl, loc: builtin definitions, usr: None), Some(Cursor(va_list kind: TypedefDecl, loc: /opt/clang+llvm-3.9.1-x86_64-linux-gnu-debian8/bin/../lib/clang/3.9.1/include/stdarg.h:30:27, usr: Some("c:@T@va_list")))
ERROR:libbindgen::ir::context: Valid declaration with no USR: Cursor(__builtin_va_list kind: TypedefDecl, loc: builtin definitions, usr: None), Some(Cursor(va_list kind: TypedefDecl, loc: /opt/clang+llvm-3.9.1-x86_64-linux-gnu-debian8/bin/../lib/clang/3.9.1/include/stdarg.h:30:27, usr: Some("c:@T@va_list")))
```

Do I miss something?

**Comment by emilio** [comment]
Those errors are non-fatal, does the generated code compile? You can post it here if you want.

If it doesn't, I bet this is #361, and you need to blacklist the functions that use varargs until we do it automatically.

**Comment by fitzgen** [comment]
We might want to make all the non-fatal `error!(...)`s into `warn!(...)`s, because I expect most people get confused. I know I did when I first started using bindgen.

Thoughts @emilio ?

**Comment by emilio** [comment]
Sounds good to me.

**Comment by sa2ajj** [comment]
right, since I had no `rpm.rs` after my previous attempts (for some reason my package installed clang 3.8 was used) I did not check whether the aforementioned file has any content: actually it does.

I tried to compile the produced `rpm.rs` and it failed: from my inexperienced look, the problems are not related to the issues that led to this issue.  A couple of errors from running rustc on the `rpm.rs`:
```
error[E0428]: a type named `poptOption` has already been defined in this module
...
error: unions are unstable and possibly buggy (see issue #32836)
...
error: aborting due to 14 previous errors
```
and the union related error is the 13 beside the first one.

**Comment by sa2ajj** [comment]
I forgot to mention:
```
$ rustc --version
rustc 1.14.0 (e8a012324 2016-12-16)
```

**Comment by emilio** [comment]
Ok, so the unions one is that by default we generate rust `union`s, which are nightly-only right now. We probably should change that default behavior.

Can you call bindgen with `--no-unstable-rust`, and see if that makes things better? The `poptOption` error seems interesting.

**Comment by sa2ajj** [comment]
As you suggested, I run `bindgen` with `--no-unstable-rust` option. The compilation of the resulting `rpm.rs` failed with
```
error[E0428]: a type named `poptOption` has already been defined in this module
    --> rpm.rs:1246:1
     |   
1176 | pub struct poptOption {
     | - previous definition of `poptOption` here
...
1246 | pub type poptOption = *mut poptOption;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ already defined

error: main function not found

error[E0391]: unsupported cyclic reference between types/traits detected
    --> rpm.rs:1246:28
     |   
1246 | pub type poptOption = *mut poptOption;
     |                            ^^^^^^^^^^ cyclic reference
     |   
     = note: the cycle begins when processing `poptOption`...
     = note: ...which then again requires processing `poptOption`, completing the cycle.

error: aborting due to previous error
```

`main` function not found is understandable, the rest is not that much.

`poptOption` comes from popt (`#include <popt.h>`) library and it has this:
```
struct poptOption {
...
};
...
#ifndef __cplusplus
/*@-exporttype -typeuse@*/
typedef struct poptOption * poptOption;
/*@=exporttype =typeuse@*/
#endif
...
```

**Comment by emilio** [comment]
Yup! passing `-x c++` to bindgen like:

```
./bindgen wrapper.h --output=rpm.rs --no-unstable-rust -- -x c++
```

Should make it work.

**Comment by emilio** [comment]
That's... Interesting!

Do you have the definition of the struct that is causing those?

**Comment by sa2ajj** [comment]
this comes from `<bits/timex.h>`, I believe, which means that `bindgen` tries to generate for all things system.

what I want is to generate bindings for rpm related headers only, is there an easy way to do that?

the code `rustc` complains about:

----
```rust
#[repr(C)]
#[derive(Debug, Copy)]
pub struct timex {
    pub modes: ::std::os::raw::c_uint,
    pub offset: __syscall_slong_t,
    pub freq: __syscall_slong_t,
    pub maxerror: __syscall_slong_t,
    pub esterror: __syscall_slong_t,
    pub status: ::std::os::raw::c_int,
    pub constant: __syscall_slong_t,
    pub precision: __syscall_slong_t,
    pub tolerance: __syscall_slong_t,
    pub time: timeval,
    pub tick: __syscall_slong_t,
    pub ppsfreq: __syscall_slong_t,
    pub jitter: __syscall_slong_t,
    pub shift: ::std::os::raw::c_int,
    pub stabil: __syscall_slong_t,
    pub jitcnt: __syscall_slong_t,
    pub calcnt: __syscall_slong_t,
    pub errcnt: __syscall_slong_t,
    pub stbcnt: __syscall_slong_t,
    pub tai: ::std::os::raw::c_int,
    pub _bitfield_1: u32,
    pub _bitfield_2: u32,
    pub _bitfield_3: u32,
    pub _bitfield_4: u32,
    pub _bitfield_5: u32,
    pub _bitfield_6: u32,
    pub _bitfield_7: u32,
    pub _bitfield_8: u32,
    pub _bitfield_9: u32,
    pub _bitfield_10: u32,
    pub _bitfield_11: u32,
}
#[test]
fn bindgen_test_layout_timex() {
    assert_eq!(::std::mem::size_of::<timex>() , 208usize);
    assert_eq!(::std::mem::align_of::<timex>() , 8usize);
}
impl Clone for timex {
    fn clone(&self) -> Self { *self }
}
impl timex {
    #[inline]
    pub fn at_offset_0(&self) -> ::std::os::raw::c_int {
        unsafe {
            ::std::mem::transmute(((self._bitfield_1 &
                                        (4294967295usize as u32)) >> 0u32) as
                                      u32)
        }
    }
    ...
    #[inline]
    pub fn at_offset_0(&self) -> ::std::os::raw::c_int {
        unsafe {
            ::std::mem::transmute(((self._bitfield_2 &
                                        (4294967295usize as u32)) >> 0u32) as
                                      u32)
        }
    }
    ...
```

**Comment by emilio** [comment]
Yes, you should whitelist the rpm stuff you want, using regexes.

For example: `--whitelist-function "rpm.*".

**Comment by fitzgen** [comment]
@sa2ajj are you still hitting this on master? If so, the best way to contribute to resolving this bug would be to [create a reduced test case](https://github.com/servo/rust-bindgen/blob/master/CONTRIBUTING.md#using-creduce-to-minimize-test-cases).

**Comment by sa2ajj** [comment]
sorry for a slow follow-up: i'll check it during the coming weekend :/

**Comment by tarcieri** [comment]
I've managed to use bindgen to produce a Rust binding for `librpm.so`, `librpmio.so`, `librpmbuild.so`, and `librpmsign.so` (heads up @sa2ajj):

https://github.com/iqlusion-io/crates/tree/master/rpmlib-sys

I should start by noting that all of the issues being discussed in this thread, at least for me, have been fixed, and IMO the issue can probably be closed.

I ran into several other issues, most of which I've managed to address. Many of them were fixed by having bindgen build a C++ binding instead of a C one. If you want the gorey details you can read about them here:

https://github.com/iqlusion-io/crates/issues/11
https://github.com/iqlusion-io/crates/issues/12

**Comment by emilio** [comment]
Ok, I'll close this based on that, thanks @tarcieri!

 * iqlusion-io/crates#11 should be fixed with rust_target(1.25), since that supports repr(align).
 * iqlusion-io/crates#12 is kind of hard to fix without introducing more complex opt-in name namespacing, but it's tracked on #1252.

**Comment by tony-iqlusion** [comment]
@emilio thanks for the pointers... looks like #1252 is about the same silly header fileI was having trouble with on https://github.com/iqlusion-io/crates/issues/12 ! Glad I found a workaround.
